mmdet/models/necks/bifpn.py

```diff
@@ -0,0 +1,327 @@
+                idx = offset - len(feature_reduction)
+                input_reduction = reduction[idx]
+
+            reduction_ratio = target_reduction / input_reduction
+
+            if used_input != output_channel:
+                conv = ConvBn(used_input, output_channel)
+                offset_nodes.add_module("conv", conv)
+
+            if reduction_ratio > 1:
+                stride_size = int(reduction_ratio)
+                offset_nodes.add_module("max_pool", nn.MaxPool2d(
+                    kernel_size=stride_size + 1, stride=stride_size, padding=1
+                ))
+
+            elif reduction_ratio < 1:
+                scale = int(1 // reduction_ratio)
+                offset_nodes.add_module("upsample", nn.UpsamplingNearest2d(
+                    scale_factor=scale))
+
+            self.input_layer[str(offset)] = offset_nodes
+
+        if self.weight_method != "sum":
+            self.edge_weights = nn.Parameter(torch.ones(len(input_offsets)), requires_grad=True)
+
+        if separable_conv:
+            self.fusion_convs = SeparableConv2d(output_channel, output_channel)
+        else:
+            self.fusion_convs = ConvBn(output_channel, output_channel)
+
+    def forward(self, inputs):
+        # Create node inputs
+        nodes = []
+        dtype = inputs[0].dtype
+
+        for offset in self.offsets:
+            nodes.append(self.input_layer[str(offset)](inputs[offset]))
+
+        for i in range(1, len(nodes)):
+            if nodes[0].size() != nodes[i].size():
+                nodes[i] = F.interpolate(nodes[i], size=(nodes[0].size(2), nodes[0].size(3)))
+
+        # Weight method
+        # Softmax normalized fusion
+        if self.weight_method == "attn":
+            normalized_weights = torch.softmax(self.edge_weights.type(dtype), dim=0)
+            nodes = torch.stack(nodes, dim=-1) * normalized_weights
+
+        # Fast normalized feature fusion
+        elif self.weight_method == "fast_attn":
+            edge_weights = F.relu(self.edge_weights.type(dtype))
+            weights_sum = torch.sum(edge_weights)
+
+            nodes = torch.stack(
+                [(nodes[i] * edge_weights[i]) / (weights_sum + self.epsilon)
+                 for i in range(len(nodes))], dim=-1)
+
+        elif self.weight_method == "sum":
+            nodes = torch.stack(nodes, dim=-1)
+
+        nodes = torch.sum(nodes, dim=-1)
+
+        nodes = self.act_layer(nodes)
+
+        return self.fusion_convs(nodes)
+
+
+class BiFPNBlock(nn.Module):
+
+    def __init__(self, input_channels, num_backbone_features, num_outs,
+                 channels, weight_method, act_fn, separable_conv, epsilon,
+                 input_offsets, feature_reduction, reduction):
+        super().__init__()
+
+        weight_method_list = ["attn", "fast_attn", "sum"]
+
+        self.num_outs = num_outs
+
+        assert self.num_outs >= 3
+
+        assert weight_method in weight_method_list
+
+        self.nodes = nn.ModuleList()
+
```

Comment:
IMO, this amount of blank lines is not reasonable.

Reply:
Is new set of blank lines in that area more reasonable?

Reply:
A bit.